File: internal/pythonfacts/object_imports.py
import hashlib
import json
import posixpath
import sys

from loader_bindings import LoaderBindings
from object_guards import object_guard
from reachability import (
    ReachabilityResolver,
    _module_object,
    _path,
    _registry_targets,
)
from type_resolver import (
    MAX_FACT_BYTES,
    MAX_MODULES,
    MAX_OUTPUT_BYTES,
    MAX_RECORD_BYTES,
    _exact,
    _unique_object,
)

LOADER = "pkgutil.resolve_name"


class ObjectImportResolver(ReachabilityResolver):
    def __init__(self, modules, root, registries):
        super().__init__(modules)
        if root != ".":
            _path(root)
        self.root = root
        self.loader_bindings = LoaderBindings(self.facts)
        self.guard_calls = {}
        self.registries = {}
        for registry in registries:
            _exact(registry, ("path", "content", "error"), "object import registry")
            _path(registry["path"])
            if registry["path"] in self.registries:
                raise ValueError("object import registry is duplicated")
            if not all(isinstance(registry[key], str) for key in ("content", "error")):
                raise TypeError("object import registry has invalid input")
            self.registries[registry["path"]] = registry

    def imports(self):
        result = []
        for path, module in sorted(self.facts.files.items()):
            for call in module["facts"]["calls"]:
                callee = call["callee"]
                if callee[
                    "kind"
                ] != "name" or LOADER not in self.loader_bindings.candidates(
                    module, call["scope"], callee["value"]
                ):
                    continue
                result.append(self.import_fact(path, module, call))
        return result

    def import_fact(self, path, module, call):
        fact = {
            "path": path,
            "callable": self.callable_name(module, call["scope"]),
            "site": call["site"],
            "endSite": call["endSite"],
            "argument": call["arguments"][0]["text"] if call["arguments"] else "",
            "targets": [],
            "registry": None,
            "guard": None,
            "error": "",
        }
        try:
            if self.callee(module, call) != LOADER:
                raise ValueError("object loader binding is ambiguous or wildcard")
            if len(call["arguments"]) != 1 or call["keywords"]:
                raise ValueError("object loader requires one positional argument")
            argument = call["arguments"][0]
            if argument["kind"] == "string":
                names = [_module_object(argument["value"])]
            elif argument["kind"] == "name":
                fact["guard"] = object_guard(self, module, call)
                names = []
            else:
                names, fact["registry"] = self.registry_targets(module, call)
            fact["targets"] = [
                {
                    "module": module_name,
                    "symbol": symbol,
                    "standardLibrary": module_name.partition(".")[0]
                    in sys.stdlib_module_names,
                }
                for module_name, symbol in sorted(set(names))
            ]
        except (ValueError, TypeError, KeyError, IndexError, RecursionError) as error:
            fact["error"] = str(error)[:4096]
        return fact

    def registry_targets(self, module, call):
        path, pointer, dynamic = self.registry_selection(module, call)
        path = posixpath.join(self.root, path) if self.root != "." else path
        registry = self.registries.get(path)
        if registry is None or registry["error"]:
            raise ValueError("object loader has no current governed registry input")
        names = _registry_targets(registry["content"], pointer, dynamic)
        return names, {
            "path": path,
            "jsonPointer": pointer,
            "sha256": hashlib.sha256(registry["content"].encode("utf-8")).hexdigest(),
        }


def resolve_stream(source, output):
    record = source.readline(MAX_RECORD_BYTES + 1)
    if len(record) > MAX_RECORD_BYTES or not record.endswith(b"\n"):
        raise ValueError("object import header exceeds its transport boundary")
    header = json.loads(record, object_pairs_hook=_unique_object)
    _exact(header, ("protocol", "count", "root", "registries"), "object import header")
    if (
        header["protocol"] != "python-object-import-project/v1"
        or type(header["count"]) is not int
        or not 0 <= header["count"] <= MAX_MODULES
        or not isinstance(header["registries"], list)
    ):
        raise ValueError("object import header is invalid")
    modules, size = [], 0
    for _ in range(header["count"]):
        record = source.readline(MAX_RECORD_BYTES + 1)
        size += len(record)
        if (
            len(record) > MAX_RECORD_BYTES
            or size > MAX_FACT_BYTES
            or not record.endswith(b"\n")
        ):
            raise ValueError("object import project exceeds its transport boundary")
        modules.append(json.loads(record, object_pairs_hook=_unique_object))
    if source.read(1):
        raise ValueError("object import project has trailing input")
    result = {
        "protocol": header["protocol"],
        "covered": [
            {"path": value["path"], "sourceSha256": value["sourceSha256"]}
            for value in modules
        ],
        "imports": ObjectImportResolver(
            modules, header["root"], header["registries"]
        ).imports(),
    }
    encoded = json.dumps(
        result, ensure_ascii=False, sort_keys=True, separators=(",", ":")
    ).encode("utf-8")
    if len(encoded) > MAX_OUTPUT_BYTES:
        raise ValueError("object import output exceeds its response boundary")
    output.write(encoded)


if __name__ == "__main__":
    resolve_stream(sys.stdin.buffer, sys.stdout.buffer)

File: internal/pythonfacts/object_imports.go
package pythonfacts

import (
	"bytes"
	"context"
	"crypto/sha256"
	_ "embed"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"path/filepath"
	"slices"
	"strings"
)

//go:embed object_imports.py
var objectImportsSource string

//go:embed object_guards.py
var objectGuardsSource string

type SourceLocation struct {
	Line   int `json:"line"`
	Column int `json:"column"`
}

type ObjectRegistryInput struct {
	Path    string `json:"path"`
	Content string `json:"content"`
	Error   string `json:"error"`
}

type ObjectRegistryEvidence struct {
	Path        string `json:"path"`
	JSONPointer string `json:"jsonPointer"`
	SHA256      string `json:"sha256"`
}

type ObjectImportTarget struct {
	Module          string `json:"module"`
	Symbol          string `json:"symbol"`
	StandardLibrary bool   `json:"standardLibrary"`
}

type ObjectInputGuard struct {
	Namespace       string                 `json:"namespace"`
	StandardLibrary bool                   `json:"standardLibrary"`
	Site            SourceLocation         `json:"site"`
	EndSite         SourceLocation         `json:"endSite"`
	Predicate       ReachabilityDefinition `json:"predicate"`
	PredicateSHA256 string                 `json:"predicateSha256"`
}

type ObjectImport struct {
	Path     string                  `json:"path"`
	Callable string                  `json:"callable"`
	Site     SourceLocation          `json:"site"`
	EndSite  SourceLocation          `json:"endSite"`
	Argument string                  `json:"argument"`
	Targets  []ObjectImportTarget    `json:"targets"`
	Registry *ObjectRegistryEvidence `json:"registry"`
	Guard    *ObjectInputGuard       `json:"guard"`
	Error    string                  `json:"error"`
}

type ObjectImportProject struct {
	Imports  []ObjectImport
	Identity string
}

func ResolveObjectImports(ctx context.Context, python, root string, modules []TypeModule, registries []ObjectRegistryInput) (ObjectImportProject, error) {
	ordered := slices.Clone(modules)
	slices.SortFunc(ordered, func(left, right TypeModule) int { return strings.Compare(left.Path, right.Path) })
	if len(ordered) > maximumProjectSources || !filepath.IsAbs(python) {
		return ObjectImportProject{}, fmt.Errorf("object import project source count or interpreter is invalid")
	}
	registries = slices.Clone(registries)
	slices.SortFunc(registries, func(left, right ObjectRegistryInput) int { return strings.Compare(left.Path, right.Path) })
	header, err := json.Marshal(struct {
		Protocol   string                `json:"protocol"`
		Count      int                   `json:"count"`
		Root       string                `json:"root"`
		Registries []ObjectRegistryInput `json:"registries"`
	}{"python-object-import-project/v1", len(ordered), root, append([]ObjectRegistryInput{}, registries...)})
	if err != nil {
		return ObjectImportProject{}, err
	}
	if len(header)+1 > maximumResponseSize {
		return ObjectImportProject{}, fmt.Errorf("object import registry header exceeds its input boundary")
	}
	input := &typeProjectReader{modules: ordered, current: bytes.NewReader(append(header, '\n'))}
	digest := sha256.New()
	program := TypeSupportSource() + ReachabilitySupportSource() + embeddedPythonModule("object_guards", objectGuardsSource) + embeddedPythonModule("__main__", objectImportsSource)
	data, err := runFactProject(ctx, python, io.TeeReader(input, digest), program)
	if err != nil {
		return ObjectImportProject{}, err
	}
	imports, err := decodeObjectImportProject(data, ordered, registries)
	if err != nil {
		return ObjectImportProject{}, err
	}
	_, _ = digest.Write(data)
	return ObjectImportProject{Imports: imports, Identity: hex.EncodeToString(digest.Sum(nil))}, nil
}

func decodeObjectImportProject(data []byte, modules []TypeModule, registries []ObjectRegistryInput) ([]ObjectImport, error) {
	var wire struct {
		Protocol string         `json:"protocol"`
		Covered  []typeCoverage `json:"covered"`
		Imports  []ObjectImport `json:"imports"`
	}
	decoder := json.NewDecoder(bytes.NewReader(data))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(&wire); err != nil {
		return nil, err
	}
	if err := decoder.Decode(&struct{}{}); err != io.EOF {
		return nil, fmt.Errorf("object import project output has trailing data")
	}
	if wire.Protocol != "python-object-import-project/v1" || wire.Imports == nil {
		return nil, fmt.Errorf("object import project output is incomplete")
	}
	files, err := projectCoveredFiles(wire.Covered, modules)
	if err != nil {
		return nil, err
	}
	if err := validateObjectImports(files, wire.Imports, registries); err != nil {
		return nil, err
	}
	if err := validateObjectImportCalls(modules, wire.Imports); err != nil {
		return nil, err
	}
	if err := validateObjectInputGuards(modules, wire.Imports); err != nil {
		return nil, err
	}
	return wire.Imports, nil
}

func validateObjectImports(files []string, imports []ObjectImport, registries []ObjectRegistryInput) error {
	seen := map[string]bool{}
	for _, imported := range imports {
		identity := fmt.Sprintf("%s:%d:%d:%d:%d", imported.Path, imported.Site.Line, imported.Site.Column, imported.EndSite.Line, imported.EndSite.Column)
		if seen[identity] || !slices.Contains(files, imported.Path) || !validObjectImportSpan(imported) {
			return fmt.Errorf("object import evidence has an invalid or duplicate callsite")
		}
		seen[identity] = true
		if err := validateObjectImportTargets(imported); err != nil {
			return err
		}
		if imported.Registry != nil {
			if err := validateObjectRegistry(*imported.Registry, registries); err != nil {
				return err
			}
		}
	}
	return nil
}

func validObjectImportSpan(value ObjectImport) bool {
	return value.Callable != "" && value.Site.Line > 0 && value.Site.Column > 0 && value.EndSite.Column > 0 &&
		(value.EndSite.Line > value.Site.Line || value.EndSite.Line == value.Site.Line && value.EndSite.Column >= value.Site.Column)
}

func validateObjectImportTargets(imported ObjectImport) error {
	if imported.Targets == nil {
		return fmt.Errorf("object import evidence omits its target collection")
	}
	if imported.Guard != nil {
		if imported.Error != "" || imported.Registry != nil || len(imported.Targets) != 0 {
			return fmt.Errorf("guarded object input has a competing outcome")
		}
		return nil
	}
	if (imported.Error == "") != (len(imported.Targets) > 0) {
		return fmt.Errorf("object import evidence has no exact outcome")
	}
	seen := map[ObjectImportTarget]bool{}
	for _, target := range imported.Targets {
		if target.Module == "" || target.Symbol == "" || seen[target] {
			return fmt.Errorf("object import evidence has an invalid or duplicate target")
		}
		seen[target] = true
	}
	return nil
}

func validateObjectRegistry(evidence ObjectRegistryEvidence, inputs []ObjectRegistryInput) error {
	for _, input := range inputs {
		if input.Path == evidence.Path {
			digest := sha256.Sum256([]byte(input.Content))
			if input.Error != "" || input.Content == "" || evidence.SHA256 != hex.EncodeToString(digest[:]) {
				return fmt.Errorf("object import registry evidence is stale or invalid")
			}
			return nil
		}
	}
	return fmt.Errorf("object import registry has no supplied input")
}
